clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+          Image.Target);
+    }
+
+    // Note: Manifests are deprecated but corresponding nullptr fields should
+    // remain to comply ABI.
+    std::pair<Constant *, Constant *> Manifests = {NullPtr, NullPtr};
+
+    // For SYCL image offload entries are defined here, by wrapper, so
+    // those are created per image
+    std::pair<Constant *, Constant *> ImageEntriesPtrs =
+        addOffloadEntriesToModule(Image.Entries);
+    Constant *WrappedImage = ConstantStruct::get(
+        SyclDeviceImageTy, Version, Kind, Format, Target, CompileOptions,
+        LinkOptions, Manifests.first, Manifests.second, Binary.first,
+        Binary.second, ImageEntriesPtrs.first, ImageEntriesPtrs.second,
+        PropSets.first, PropSets.second);
+
+    if (Options.EmitRegistrationFunctions)
+      emitRegistrationFunctions(Binary.first, Image.Image.size(), ImageID,
+                                OffloadKindTag);
+
+    return WrappedImage;
+  }
+
+  GlobalVariable *combineWrappedImages(ArrayRef<Constant *> WrappedImages,
+                                       StringRef OffloadKindTag) {
+    auto *ImagesData = ConstantArray::get(
+        ArrayType::get(SyclDeviceImageTy, WrappedImages.size()), WrappedImages);
+    auto *ImagesGV =
+        new GlobalVariable(M, ImagesData->getType(), /*isConstant*/ true,
+                           GlobalValue::InternalLinkage, ImagesData,
+                           Twine(OffloadKindTag) + "device_images");
+    ImagesGV->setUnnamedAddr(GlobalValue::UnnamedAddr::Global);
+
+    auto *Zero = ConstantInt::get(getSizeTTy(), 0);
+    Constant *ZeroZero[] = {Zero, Zero};
+    auto *ImagesB = ConstantExpr::getGetElementPtr(ImagesGV->getValueType(),
+                                                   ImagesGV, ZeroZero);
+
+    // And finally create the binary descriptor object.
+    Constant *EntriesB = Constant::getNullValue(PointerType::getUnqual(C));
+    Constant *EntriesE = Constant::getNullValue(PointerType::getUnqual(C));
+    static constexpr uint16_t BinDescStructVersion = 1;
+    auto *DescInit = ConstantStruct::get(
+        getSyclBinDescTy(),
+        ConstantInt::get(Type::getInt16Ty(C), BinDescStructVersion),
+        ConstantInt::get(Type::getInt16Ty(C), WrappedImages.size()), ImagesB,
+        EntriesB, EntriesE);
+
+    return new GlobalVariable(M, DescInit->getType(), /*isConstant*/ true,
+                              GlobalValue::InternalLinkage, DescInit,
+                              Twine(OffloadKindTag) + Twine("descriptor"));
+  }
+
+  /// Creates binary descriptor for the given device images. Binary descriptor
+  /// is an object that is passed to the offloading runtime at program startup
+  /// and it describes all device images available in the executable or shared
+  /// library. It is defined as follows
+  ///
+  ///
+  /// __attribute__((visibility("hidden")))
+  /// extern __tgt_offload_entry *__start_offloading_entries0;
```

Comment:
Why 0 at the end? Note that this is a one more conflict with the upstream code

Reply:
0 here means an index of entries from the first image (with index 0).
There are following `...` that are supposed to expand to:
```
__attribute__((visibility("hidden")))
extern __tgt_offload_entry *__start_offloading_entries0;
__attribute__((visibility("hidden")))
extern __tgt_offload_entry *__stop_offloading_entries0;
__attribute__((visibility("hidden")))
extern __tgt_offload_entry *__start_offloading_entries1;
__attribute__((visibility("hidden")))
extern __tgt_offload_entry *__stop_offloading_entries1;
...
__attribute__((visibility("hidden")))
extern __tgt_offload_entry *__start_offloading_entriesN;
__attribute__((visibility("hidden")))
extern __tgt_offload_entry *__stop_offloading_entriesN;
```

Also the first image is supposed to be:
```
{
 Version,                      /*Version*/
 OffloadKind,                  // Kind of offload model.
 Format,                       // format of the image - SPIRV, LLVMIR
                               // bc, etc
 NULL,                         /*DeviceTargetSpec*/
 CompileOptions0,              /*CompileOptions0*/
 LinkOptions0,                 /*LinkOptions0*/
 NULL,                         /*ManifestStart*/
 NULL,                         /*ManifestEnd*/
 Image0,                       /*ImageStart*/
 Image0 + sizeof(Image0),      /*ImageEnd*/
 __start_offloading_entries0,  /*EntriesBegin*/
 __stop_offloading_entries0,   /*EntriesEnd*/ 
 PropertySetBegin0,            /*EntriesEnd*/
 PropertySetEnd0               /*EntriesEnd*/
}
```